src/librustc_data_structures/vob_map_set.rs

```diff
@@ -0,0 +1,247 @@
+use vob::{self, Vob};
+use std::iter::Iterator;
+
+/// Conceptually this is a `Map<usize, Set<usize>>`, using a single bit vector internally.
+///
+/// * The size of the sets for each key must be the same and known ahead of time.
+/// * Operations will panic if you try to access out of bounds keys or set items.
+/// * When the map is resized, new keys will be instantiated with empty sets. As such, this data
+///   structure cannot reliably express the absence of a key.
+pub struct VobMapSet {
+    vob: Vob,
+    set_size: usize,
+}
+
+impl VobMapSet {
+    /// Create a new `VobMapSet` with `initial_num_keys` keys, whose sets are empty.
+    pub fn new(initial_num_keys: usize, set_size: usize) -> Self {
+        Self {
+            set_size,
+            vob: Vob::from_elem(set_size.checked_mul(initial_num_keys).unwrap(), false),
+        }
+    }
+
+    /// Returns the bit index for the bit for `val` in `key`.
+    fn idx(set_size: usize, key: usize, val: usize) -> usize {
+        key.checked_mul(set_size).unwrap().checked_add(val).unwrap()
+    }
+
+    /// Ensure there is space for `num_keys` keys.
+    /// If `num_keys` is smaller than `self.len()` the map is truncated.
+    /// If `num_keys` is larger, then new keys are allocated with empty sets.
+    pub fn resize(&mut self, num_keys: usize) {
+        let required_size = num_keys.checked_mul(self.set_size).unwrap();
+        self.vob.resize(required_size, false);
+    }
+
+    /// Insert `val` into the set for key `key`.
+    pub fn insert(&mut self, key: usize, val: usize) {
+        self.vob.set(Self::idx(self.set_size, key, val), true);
+    }
+
+    /// Insert `val` into the set for key `key`, resizing the map if necessary.
+    pub fn insert_maybe_resize(&mut self, key: usize, val: usize) {
+        let required_size = key.checked_add(1).unwrap().checked_mul(self.set_size).unwrap();
+        if self.vob.len() < required_size {
+            self.vob.resize(required_size, false);
+        }
+        self.insert(key, val);
+    }
+
+    /// Insert `val` into the set for key `key`. `v` must not be larger than the set size.
+    pub fn contains(&mut self, key: usize, val: usize) -> bool {
+        self.vob.get(Self::idx(self.set_size, key, val)).unwrap()
+    }
+
+    pub fn pop(&mut self, key: usize) -> Option<usize> {
+        let start = Self::idx(self.set_size, key, 0);
```

Comment:
Isn't this roughly equivalent to something like:

```rust
self.vob.iter_set_bits(self.set_size * ..., self.set_size * ...).next().unwrap_or(None)
```

?

Reply:
Ah ha! Yes, that'd be neater.